src/resolver.ts

```diff
@@ -11,35 +16,60 @@ async function get(url: string): Promise<any> {
   return res.json()
 }
 
-export function getResolver() {
+export function getResolver(): Record<string, DIDResolver> {
   async function resolve(
     did: string,
     parsed: ParsedDID
-  ): Promise<DIDDocument | null> {
-    let path = parsed.id + DOC_PATH
+  ): Promise<DIDResolutionResult> {
+    let err = null
+    let path = decodeURIComponent(parsed.id) + DOC_PATH
     const id = parsed.id.split(':')
-    if (id.length > 1) path = id.join('/') + '/did.json'
+    if (id.length > 1) {
+      path = id.map(decodeURIComponent).join('/') + '/did.json'
+    }
+
     const url: string = `https://${path}`
 
-    let data: any = null
-    try {
-      data = await get(url)
-    } catch (error) {
-      throw new Error(
-        `DID must resolve to a valid https URL containing a JSON document: ${error.message}`
-      )
-    }
+    const didDocumentMetadata = {}
+    let didDocument: DIDDocument | null = null
 
-    const docIdMatchesDid = data.id === did
-    if (!docIdMatchesDid) {
-      throw new Error('DID document id does not match requested did')
-    }
+    do {
+      try {
+        didDocument = await get(url)
+      } catch (error) {
+        err = `DID must resolve to a valid https URL containing a JSON document: ${error}`
+        break
+      }
+
+      // TODO: this excludes the use of query params
+      const docIdMatchesDid = didDocument?.id === did
+      if (!docIdMatchesDid) {
+        err = 'DID document id does not match requested did'
+        break
```

Comment:
This break can be removed, correct?

Reply:
since it's the last one, yes it can be removed. I kept it here in case the list of checks needs to grow.
In previous versions we had a check to see if the returned DID documents contained any public keys
```suggestion
```